pravega-video/src/index.rs

```diff
@@ -269,42 +278,49 @@ impl<R: Read + Seek + CurrentHead> IndexSearcher<R> {
+                return Ok((last_index_record, last_index_offset));
             }
 
             // Read first record.
-            let mut prev_index_offset = self.reader.seek(SeekFrom::Start(first_index_offset))?;
-            let mut prev_index_record = index_record_reader.read(&mut self.reader)?;
+            let mut first_index_offset = self.reader.seek(SeekFrom::Start(first_index_offset))?;
+            let mut first_index_record = index_record_reader.read(&mut self.reader)?;
             // Return first record if desired timestamp is before or equal to it.
-            if timestamp <= prev_index_record.timestamp {
-                return Ok((prev_index_record, prev_index_offset));
+            if timestamp <= first_index_record.timestamp {
+                return Ok((first_index_record, first_index_offset));
             }
 
-            // Read index from the beginning until we find a timestamp greater or equal to the desired timestamp.
-            // TODO: Use binary search or Newton's method.
             loop {
-                let index_record = index_record_reader.read(&mut self.reader)?;
-                let index_offset = prev_index_offset + IndexRecord::RECORD_SIZE as u64;
-                trace!("IndexSearcher::search_timestamp_and_return_index_offset: index_record={:?}", index_record);
-                if timestamp < index_record.timestamp {
-                    // Approximate match returns index record before or after desired timestamp, depending on method.
-                    return match method {
-                        SearchMethod::Before => Ok((prev_index_record, prev_index_offset)),
-                        SearchMethod::After => Ok((index_record, index_offset)),
-                    }
-                } else if timestamp == index_record.timestamp {
-                    // Exact match
-                    return Ok((index_record, index_offset));
+                let middle_index = (last_index_offset + first_index_offset) / 2 / IndexRecord::RECORD_SIZE as u64;
+                let middle_index_offset = self.reader.seek(SeekFrom::Start(middle_index * IndexRecord::RECORD_SIZE as u64))?;
+                let middle_index_record = index_record_reader.read(&mut self.reader)?;
+                if timestamp < middle_index_record.timestamp {
+                    last_index_offset = middle_index_offset - IndexRecord::RECORD_SIZE as u64;
+                } else if timestamp > middle_index_record.timestamp {
+                    first_index_offset = middle_index_offset + IndexRecord::RECORD_SIZE as u64;
+                } else {
+                    return Ok((middle_index_record, middle_index_offset));
+                }
+                if first_index_offset > last_index_offset {
+                    break;
                 }
-                prev_index_offset = index_offset;
-                prev_index_record = index_record;
+            }
+            // first_index_offset == last_index_offset + IndexRecord::RECORD_SIZE
```

Comment:
Is this commented code or are you explaining an assumption/constraint? Consider making this an assert statement.

Reply:
This is the condition when break from loop. Mark as a comment. Deleted in the new commit.